# Hypothesis Testing
------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------------------

# completion rates KPI

To determine if the difference in completion rates between the new design (Test) and the old design (Control) is statistically significant, we can conduct a hypothesis test. Specifically, we'll use a two-proportion z-test.



## Load Dataset:

In [12]:
import pandas as pd
import numpy as np
from statsmodels.stats.proportion import proportions_ztest


In [36]:
completion_rate_df = pd.read_csv('../Data/completion_rates.csv')
completion_rate_df

Variation  confirm_count  total_users  completion_rate
0   Control           2879         8252         0.348885
1      Test           3119         8535         0.365436

The steps for Hypothesis Testing are as follows:

## Define the hypotheses:

1- Null hypothesis ($H_0$): The completion rates for the Test and Control groups are the same.

2- Alternative hypothesis ($H_1$): The completion rate for the Test group is greater than that for the Control group.

## Calculate the Sample Proportions:

- $p_1$= completion rate of the Test group.

- $p_2$= completion rate of the Control group.

In [4]:
# Completion rates data
completion_test = completion_rate_df[completion_rate_df['Variation'] == 'Test']
completion_control = completion_rate_df[completion_rate_df['Variation'] == 'Control']

## Calculate the Pooled Proportion and Standard Error:

- Pooled Proportion $p_{pool}$:

$p_{pool} = \frac{X_1 + X_2}{n_1 + n_2}$

Where $X_1$ and $X_2$ are the number of successes (completions) in the Test and Control groups, and $n_1$ and $n_2$ are the sample sizes.

Standard Error $SE$:

$SE = \sqrt{p_{pool} * (1 - p_{pool} ) * (1/n_1 + 1/n_2)}$

In [8]:
# Number of completions and total users in each group
X1 = completion_test['confirm_count'].values[0]
n1 = completion_test['total_users'].values[0]
X2 = completion_control['confirm_count'].values[0]
n2 = completion_control['total_users'].values[0]

print(f"X1: {X1}     n1: {n1}     X2: {X2}     n2: {n2}")

X1: 3119     n1: 8535     X2: 2879     n2: 8252


## Calculate the Z-score:

$Z = \frac{p_1 - p_2}{SE}$

## Calculate the P-value:

Using the Z-score, we can determine the P-value to see if the observed difference is statistically significant.

In [14]:
# Calculate the z-score and p-value
count = np.array([X1, X2])
nobs = np.array([n1, n2])
z_score, p_value = proportions_ztest(count, nobs, alternative='two-sided')

z_score, p_value

(2.2372097886733453, 0.025272631894062626)

## Compare with Significance Level:

Typically, a $5 $% significance level ($α = 0.05$) is used. If P-value $< 0.05$, we reject the null hypothesis. So the  P-value is about $0.02$. We **reject the null hypothesis** and conclude that **there is a statistically significant difference in completion rates between the Test and Control groups**. The observed increase in completion rate from the A/B test meets or exceeds the 5% threshold for significance.

------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------------------

# Carried out another hypothesis test of your choosing

We try to do the same hypothesis Testing for Error rate data

## Load Dataset:

In [35]:
error_rate_df = pd.read_csv('../Data/error_rates.csv')
error_rate_df

Variation  total_steps  error_steps  error_rate
0   Control        23532         2272    0.096549
1      Test        26968         2647    0.098153

The steps for Hypothesis Testing are as follows:

## Define the hypotheses:

1- Null hypothesis ($H_0$): The error rates for the Test and Control groups are the same.

2- Alternative hypothesis ($H_1$): The error rate for the Test group is greater than that for the Control group.

## Calculate the Sample Proportions:

- $p_1$= error rate of the Test group.

- $p_2$= error rate of the Control group.

In [37]:
# Completion rates data
error_test = error_rate_df[error_rate_df['Variation'] == 'Test']
error_control = error_rate_df[error_rate_df['Variation'] == 'Control']

## Calculate the Pooled Proportion and Standard Error:

- Pooled Proportion $p_{pool}$:

$p_{pool} = \frac{X_1 + X_2}{n_1 + n_2}$

Where $X_1$ and $X_2$ are the number of errors (backwarding in steps) in the Test and Control groups, and $n_1$ and $n_2$ are the sample sizes.

Standard Error $SE$:

$SE = \sqrt{p_{pool} * (1 - p_{pool} ) * (1/n_1 + 1/n_2)}$

In [39]:
# Number of completions and total users in each group
X1 = error_test['error_steps'].values[0]
n1 = error_test['total_steps'].values[0]
X2 = error_control['error_steps'].values[0]
n2 = error_control['total_steps'].values[0]

print(f"X1: {X1}     n1: {n1}     X2: {X2}     n2: {n2}")

X1: 2647     n1: 26968     X2: 2272     n2: 23532


## Calculate the Z-score:

$Z = \frac{p_1 - p_2}{SE}$

## Calculate the P-value:

Using the Z-score, we can determine the P-value to see if the observed difference is statistically significant.

In [40]:
# Calculate the z-score and p-value
count = np.array([X1, X2])
nobs = np.array([n1, n2])
z_score, p_value = proportions_ztest(count, nobs, alternative='two-sided')

z_score, p_value

(0.6064152178981506, 0.5442390656402112)

## Compare with Significance Level:

Typically, a $5 $% significance level ($α = 0.05$) is used. If P-value $< 0.05$, we reject the null hypothesis. So the  P-value is about $0.54$. We **accept the null hypothesis** and conclude that **there is not a statistically significant difference in error rates between the Test and Control groups**. The observed increase in error rate from the A/B test meets or exceeds the 5% threshold for significance.

------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------------------